### ISSUE TITLE
Issues with pom packaged modules in multi module build

### ISSUE THREAD

**Issue by bruceskingle** [symphonyoss/ssf-parent-pom#3]
I now can't reproduce all the failure scenarios I saw earlier, but here are the instructions to see what I can share.

```
$ mkdir /tmp/test
$ cd /tmp/test
$ git clone https://github.com/symphonyoss/symphony-rest-tools.git
Cloning into 'symphony-rest-tools'...
remote: Counting objects: 134, done.
remote: Compressing objects: 100% (58/58), done.
remote: Total 134 (delta 27), reused 61 (delta 5), pack-reused 51
Receiving objects: 100% (134/134), 50.08 KiB | 0 bytes/s, done.
Resolving deltas: 100% (39/39), done.
$ cd symphony-rest-tools
$ mvn clean install
[INFO] Scanning for projects...
Downloading: https://oss.sonatype.org/content/groups/public/org/symphonyoss/symphonyoss/6/symphonyoss-6.pom
Downloaded: https://oss.sonatype.org/content/groups/public/org/symphonyoss/symphonyoss/6/symphonyoss-6.pom (19 KB at 2.9 KB/sec)
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building Symphony REST tools 0.1.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
**<SNIP>**
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 18.624 s
[INFO] Finished at: 2017-09-17T16:01:11-07:00
[INFO] Final Memory: 37M/597M
[INFO] ------------------------------------------------------------------------
$ cd ..
$ git clone https://github.com/symphonyoss/symphony-rest-tools-ui.git
Cloning into 'symphony-rest-tools-ui'...
remote: Counting objects: 106, done.
remote: Compressing objects: 100% (56/56), done.
remote: Total 106 (delta 30), reused 97 (delta 24), pack-reused 0
Receiving objects: 100% (106/106), 27.75 KiB | 0 bytes/s, done.
Resolving deltas: 100% (30/30), done.
$ cd symphony-rest-tools-ui/
$ mvn clean install
[INFO] Scanning for projects...
**<SNIP>**
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Build Order:
[INFO] 
[INFO] symphony-rest-tools-ui
[INFO] bundles
[INFO] org.symphonyoss.symphony.tools.rest.ui.dependencies
[INFO] org.symphonyoss.symphony.tools.rest.ui.console
[INFO] org.symphonyoss.symphony.tools.rest.ui
[INFO] features
[INFO] org.symphonyoss.symphony.tools.rest.ui.feature
[INFO] org.symphonyoss.symphony.tools.rest.ui.console.feature
[INFO] products
[INFO] update
[INFO] org.symphonyoss.symphony.tools.rest.ui.product
**<SNIP>**
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] symphony-rest-tools-ui ............................. SUCCESS [  0.162 s]
[INFO] bundles ............................................ SUCCESS [  0.005 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.dependencies  SUCCESS [  0.926 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.console ..... SUCCESS [  0.606 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui ............. SUCCESS [  0.190 s]
[INFO] features ........................................... SUCCESS [  0.006 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.feature ..... SUCCESS [  0.078 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.console.feature SUCCESS [  0.068 s]
[INFO] products ........................................... SUCCESS [  0.005 s]
[INFO] update ............................................. SUCCESS [  1.605 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.product ..... SUCCESS [ 19.220 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 29.239 s
[INFO] Finished at: 2017-09-17T16:02:12-07:00
[INFO] Final Memory: 98M/1069M
[INFO] ------------------------------------------------------------------------
$ 
```

So far so good, we need the S2-super-pom now

```
$ cd ..
$ git clone https://github.com/SymphonyOSF/S2-super-pom.git
Cloning into 'S2-super-pom'...
remote: Counting objects: 21, done.
remote: Compressing objects: 100% (19/19), done.
remote: Total 21 (delta 8), reused 12 (delta 1), pack-reused 0
Unpacking objects: 100% (21/21), done.
$ cd S2-super-pom/
$ mvn clean install
[INFO] Scanning for projects...
**<SNIP>**
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 3.265 s
[INFO] Finished at: 2017-09-17T16:07:52-07:00
[INFO] Final Memory: 16M/207M
[INFO] ------------------------------------------------------------------------
$ cd ..
$ cd symphony-rest-tools-ui/
```

Now edit pom.xml and uncomment the parent pom entry

```
        <!-- <parent>
                <groupId>org.symphonyoss.s2</groupId>
                <artifactId>S2-super-pom</artifactId>
                <version>1.0.1-SNAPSHOT</version>
        </parent> -->
```

Now building results in this error:
```
$ mvn clean install
**<SNIP>**
[INFO] ------------------------------------------------------------------------
[INFO] Building update 1.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:3.0.0:clean (default-clean) @ update ---
[INFO] Deleting /private/tmp/test/symphony-rest-tools-ui/products/update/target
[INFO] 
[INFO] --- tycho-packaging-plugin:1.0.0:build-qualifier-aggregator (default-build-qualifier-aggregator) @ update ---
[INFO] The project's OSGi version is 1.0.0.201709172309
[INFO] 
[INFO] --- maven-clean-plugin:3.0.0:clean (default-clean-1) @ update ---
[INFO] 
[INFO] --- maven-remote-resources-plugin:1.5:process (process-resource-bundles) @ update ---
/private/tmp/test/symphony-rest-tools-ui/features/org.symphonyoss.symphony.tools.rest.ui.console.feature/.polyglot.build.properties
/private/tmp/test/symphony-rest-tools-ui/features/org.symphonyoss.symphony.tools.rest.ui.feature/.polyglot.build.properties
[INFO] 
[INFO] --- maven-resources-plugin:3.0.2:resources (default-resources) @ update ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /private/tmp/test/symphony-rest-tools-ui/products/update/src/main/resources
[INFO] Copying 3 resources
[INFO] 
[INFO] --- target-platform-configuration:1.0.0:target-platform (default-target-platform) @ update ---
[INFO] 
[INFO] --- tycho-p2-publisher-plugin:1.0.0:publish-products (default-publish-products) @ update ---
[INFO] 
[INFO] --- tycho-p2-publisher-plugin:1.0.0:publish-categories (default-publish-categories) @ update ---
[INFO] 
[INFO] --- tycho-p2-publisher-plugin:1.0.0:attach-artifacts (default-attach-artifacts) @ update ---
[INFO] 
[INFO] --- tycho-p2-repository-plugin:1.0.0:assemble-repository (default-assemble-repository) @ update ---
[INFO] Copying resources from /private/tmp/test/symphony-rest-tools-ui/products/update/target/classes to /private/tmp/test/symphony-rest-tools-ui/products/update/target/repository
[INFO] 
[INFO] --- tycho-p2-repository-plugin:1.0.0:archive-repository (default-archive-repository) @ update ---
[INFO] Building zip: /private/tmp/test/symphony-rest-tools-ui/products/update/target/update-1.0.0-SNAPSHOT.zip
[INFO] 
[INFO] >>> maven-source-plugin:3.0.0:jar (default) > generate-sources @ update >>>
[INFO] 
[INFO] --- tycho-packaging-plugin:1.0.0:build-qualifier-aggregator (default-build-qualifier-aggregator) @ update ---
[INFO] The project's OSGi version is 1.0.0.201709172309
[INFO] 
[INFO] --- maven-clean-plugin:3.0.0:clean (default-clean-1) @ update ---
[INFO] Deleting /private/tmp/test/symphony-rest-tools-ui/products/update/target
[INFO] 
[INFO] <<< maven-source-plugin:3.0.0:jar (default) < generate-sources @ update <<<
[INFO] 
[INFO] --- maven-source-plugin:3.0.0:jar (default) @ update ---
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO] 
[INFO] symphony-rest-tools-ui ............................. SUCCESS [  0.742 s]
[INFO] bundles ............................................ SUCCESS [  0.024 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.dependencies  SUCCESS [  1.997 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.console ..... SUCCESS [  0.749 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui ............. SUCCESS [  0.483 s]
[INFO] features ........................................... SUCCESS [  0.034 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.feature ..... SUCCESS [  0.174 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.console.feature SUCCESS [  0.181 s]
[INFO] products ........................................... SUCCESS [  0.026 s]
[INFO] update ............................................. FAILURE [  2.964 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.product ..... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 13.983 s
[INFO] Finished at: 2017-09-17T16:09:54-07:00
[INFO] Final Memory: 78M/1013M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-source-plugin:3.0.0:jar (default) on project update: Error adding directory to source archive. /private/tmp/test/symphony-rest-tools-ui/products/update/target/maven-shared-archive-resources isn't a directory. -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :update
$ 
```

This error was previously happening in bundles and features too, now it seems to work and I have no idea why. Also it was the case that if you delete the part of bundles/pom.xml between

```
<!-- Should not be needed START -->

and

<!-- Should not be needed END -->
```

the build would fail, but now it works, again I don't understand why.

**Comment by pmonks** [comment]
Probably not helpful, but I tried the commands listed above here, and it worked for me too, even after installing the `S2-super-pom` (which I was surprised worked, given I didn't think I had access to the SymphonyOSF GitHub organisation).

**Comment by bruceskingle** [comment]
I made the repos we are trying to open source public so you can see S2-super-pom and S2-common.

Did you get and error after uncommenting the parent in symphony-rest-tools-ui/pom.xml?

**Comment by pmonks** [comment]
Yep:

```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] symphony-rest-tools-ui ............................. SUCCESS [  2.843 s]
[INFO] bundles ............................................ SUCCESS [  0.023 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.dependencies  SUCCESS [  3.158 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.console ..... SUCCESS [  0.989 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui ............. SUCCESS [  0.405 s]
[INFO] features ........................................... SUCCESS [  0.017 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.feature ..... SUCCESS [  0.141 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.console.feature SUCCESS [  0.582 s]
[INFO] products ........................................... SUCCESS [  0.017 s]
[INFO] update ............................................. FAILURE [  1.942 s]
[INFO] org.symphonyoss.symphony.tools.rest.ui.product ..... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 18.530 s
[INFO] Finished at: 2017-09-17T17:52:37-07:00
[INFO] Final Memory: 75M/857M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-source-plugin:3.0.0:jar (default) on project update: Error adding directory to source archive.: /Users/pmonks/Development/symphony/symphony-rest-tools-ui/products/update/target/maven-shared-archive-resources isn't a directory. -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR]
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :update
```

**Comment by maoo** [comment]
Thanks for the issue. I could reproduce the problem, which is caused by the https://github.com/symphonyoss/symphony-rest-tools-ui/blob/master/products/update/pom.xml (and similar) module, where the source-plugin fails because the `target` folder is not created during the build. For these modules, we need to skip the execution of the `maven-source-plugin`.

By running `mvn package -Dmaven.source.skip=true` I get another build issue, which makes me think that it fixes the current issue; @bruceskingle - could you please try this build command locally? If your build succeeds, I can make this parameter configurable and add the proper documentation in the `README.md`.

An alternative approach is to move the `maven-source-plugin` logic into a dedicated profile (ie `-Pattach-sources`), so that it doesn't get enabled in the default build lifecycle. I personally prefer the former solution, as it doesn't have any impact on the projects that are currently using this parent POM.

**Comment by maoo** [comment]
I've analysed further the issue... it seems that the `maven-install-plugin` ignores that the source-plugin is skipped, as it logs the following (right before failing):

```
[INFO] No primary artifact to install, installing attached artifacts instead.
```

The only short term fix for the issue is to isolate the maven-source-plugin logic into a dedicated profile called `-Pattach-sources`; I'll make a final test as soon as I deploy a SNAPSHOT version of the Parent POM.

In parallel, I'll try to fix the issue on the maven-install-plugin, shouldn't be very hard.

**Comment by maoo** [comment]
@bruceskingle - could you please using the following parent pom (on `S2-super-pom` project) and try again?

```
<parent>
	<groupId>org.symphonyoss</groupId>
	<artifactId>symphonyoss</artifactId>
	<version>8-SNAPSHOT</version>
</parent>
```

**Comment by bruceskingle** [comment]
8-SNAPSHOT does not resolve and 7 and 7-SNAPSHOT don't fix the problem. Note that S2-super-pom points to the LLC artefactory repo, somehow your 7 and 7-SNAPSHOT versions are in there but 8 isn't.

Do I need to add your snapshot repo to the pom?

[ERROR] The build could not read 1 project -> [Help 1]
[ERROR]   
[ERROR]   The project org.symphonyoss.s2:S2-super-pom:1.0.1-SNAPSHOT (/Users/bruce.skingle/symphony/git/S2-super-pom/pom.xml) has 1 error
[ERROR]     Non-resolvable parent POM for org.symphonyoss.s2:S2-super-pom:1.0.1-SNAPSHOT: Could not find artifact org.symphonyoss:symphonyoss:pom:8-SNAPSHOT in snapshots (https://repo.symphony.com/artifactory/libs-snapshot) and 'parent.relativePath' points at wrong local POM @ line 36, column 11 -> [Help 2]

**Comment by bruceskingle** [comment]
Is it possible to control the activation of that plugin based on the pom packaging type?

**Comment by maoo** [comment]
> Is it possible to control the activation of that plugin based on the pom packaging type?

Unfortunately I couldn't find a way to configure this activation, the only similar config is `skipSource`, see https://maven.apache.org/plugins/maven-source-plugin/jar-no-fork-mojo.html, though it leads to the issue on `maven-install-plugin`, as documented above.

Re. the test of `8-SNAPSHOT` version, I forgot to ask you to add the following repository in your pom.xml (this won't be necessary as soon as I release version 8 on Maven Central), which is why you couldn't resolve it.

```
  <repositories>
        <repository>
            <id>sonatype-oss-public</id>
            <url>https://oss.sonatype.org/content/groups/public/</url>
            <releases>
                <enabled>true</enabled>
            </releases>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
  </repositories>
```

I can see the version `8-SNAPSHOT` in the repository - https://oss.sonatype.org/#nexus-search;gav~org.symphonyoss~symphonyoss~~~

I'm quite confident this will fix the current issue.

**Comment by bruceskingle** [comment]
Confirmed fixed, I haven't looked into what's happening in there but my build now works and I can remove the developers and other redundant sections in my aggregator poms, I have no idea why that problem came and then went.

I have just pushed a big set of commits including rewinding the versions on S2-common and S2-super-pom from 1.0.X to 0.1.X so if you have been building these I strongly suggest you clear out your local maven cache to avoid nasty surprises.

Thanks for your help with this.
